# Basic Prerequisite check :-

In [ ]:
#Checkin the Currunt Working Directry
import os
print(os.getcwd())

# Importing the necessary libraries

In [2]:
import cv2
import numpy as np
import matplotlib

print(cv2.__version__)

4.1.2


# Reading, Writing Images

In [3]:
#Reading the image
input_image = cv2.imread('input_1.jpg')

#Showing the image
cv2.imshow('Taj Mahal Image', input_image)

#Wait for the any key input for closing the input except 0 By placing numbers (except 0), we can specify a delay for
# how long you keep the window open (time is in milliseconds here)
cv2.waitKey()

# This closes all open windows 
# Failure to place this will cause your program to hang
cv2.destroyAllWindows()

# Printing Shapes of Images

In [7]:
print(input_image.shape)

#Shapes is just a pixel height * Width * depth of image (Depth=3 because image is in RGB)

(340, 527, 3)


In [8]:
print("Height of image is ",input_image.shape[0])
print("Width of image is ",input_image.shape[1])

Height of image is  340
Width of image is  527


# Grey Scaling Of Images

1.https://docs.opencv.org/2.4/modules/imgproc/doc/miscellaneous_transformations.html

In [9]:
#Reading the image
input_image = cv2.imread('input_1.jpg')
#Showing the image
cv2.imshow('Taj Mahal Image', input_image)
cv2.waitKey()


#Converting the image to gray scale
gray_image = cv2.cvtColor(input_image, cv2.COLOR_BGR2GRAY)
cv2.imshow('Grayscale', gray_image)
cv2.waitKey()


cv2.destroyAllWindows()

In [2]:
#Second methos for converting image to gray scale
input_image = cv2.imread('input_1.jpg',0)

cv2.imshow('Taj Mahal Image Grayscale', input_image)
cv2.waitKey()
cv2.destroyAllWindows()

# Introduction to color spaces

https://opencv-python-tutroals.readthedocs.io/en/latest/py_tutorials/py_imgproc/py_colorspaces/py_colorspaces.html

In [3]:
input_image = cv2.imread('input_1.jpg')
#Showing the image
cv2.imshow('Taj Mahal Image', input_image)
cv2.waitKey()

# BGR Values for the first 0,0 pixel
B, G, R = input_image[0, 0] 
print(B, G, R)
print(input_image.shape)

cv2.destroyAllWindows()


229 174 131
(340, 527, 3)


From Above output we can say that the First pixels contans 230 as blue + 174 as green and 133 as red pixels.

In [4]:
gray_img = cv2.cvtColor(input_image, cv2.COLOR_BGR2GRAY)
print(gray_img.shape)
print(gray_img[10, 50]) 

(340, 527)
168


# Drawing on empty Images

https://docs.opencv.org/2.4/modules/core/doc/drawing_functions.html

In [6]:
# Creating an empty image
image = np.zeros((512,512,3), np.uint8)

#Drawing line on empty image
cv2.line(image, (0,0), (511,511), (255,127,0), 5)
cv2.imshow("Blue Line", image)

cv2.waitKey(0)
cv2.destroyAllWindows()

In [7]:
image = np.zeros((512,512,3), np.uint8)

#Drawing Circle on Empty Image
cv2.circle(image, (350, 350), 100, (15,75,50), -1) 
cv2.imshow("Circle", image)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [10]:
image = np.zeros((512,512,3), np.uint8)

#Putting Text on an empty image
cv2.putText(image, 'Hello World!', (75,290), cv2.FONT_HERSHEY_SCRIPT_COMPLEX, 2, (100,170,0), 3)
cv2.imshow("Hello World!", image)
cv2.waitKey(0)
cv2.destroyAllWindows()


# Sharpening Of an Image

https://docs.opencv.org/2.4/modules/imgproc/doc/filtering.html

In [11]:
import cv2
import numpy as np

image =  cv2.imread('input_1.jpg')
cv2.imshow('Original', image)

# Create our shapening kernel, we don't normalize since the 
# the values in the matrix sum to 1
kernel_sharpening = np.array([[-1,-1,-1], 
                              [-1,9,-1], 
                              [-1,-1,-1]])

# applying different kernels to the input image
sharpened = cv2.filter2D(image, -1, kernel_sharpening)

cv2.imshow('Image Sharpening', sharpened)

cv2.waitKey(0)
cv2.destroyAllWindows()

# Edge Detection
https://opencv-python-tutroals.readthedocs.io/en/latest/py_tutorials/py_imgproc/py_gradients/py_gradients.html
https://opencv-python-tutroals.readthedocs.io/en/latest/py_tutorials/py_imgproc/py_canny/py_canny.html

In [4]:
import cv2
import numpy as np

#reading the image
image =  cv2.imread('input_1.jpg',0)

height, width = image.shape

# Extract Sobel Edges
sobel_x = cv2.Sobel(image, cv2.CV_64F, 0, 1, ksize=5)
sobel_y = cv2.Sobel(image, cv2.CV_64F, 1, 0, ksize=5)

cv2.imshow('Original', image)
cv2.waitKey(0)
cv2.imshow('Sobel X', sobel_x)
cv2.waitKey(0)
cv2.imshow('Sobel Y', sobel_y)
cv2.waitKey(0)


#Printing the sobel edges for both x and y axis
sobel_OR = cv2.bitwise_or(sobel_x, sobel_y)
cv2.imshow('sobel_OR', sobel_OR)
cv2.waitKey(0)


laplacian = cv2.Laplacian(image, cv2.CV_64F)
cv2.imshow('Laplacian', laplacian)
cv2.waitKey(0)


#Detecting the canny edges
canny = cv2.Canny(image,50, 150)
cv2.imshow('Canny', canny)
cv2.waitKey(0)

canny = cv2.Canny(image,50, 150)
cv2.imshow('Canny', canny)
cv2.waitKey(0)

cv2.destroyAllWindows()

# Live Sketch App

In [ ]:
import cv2
import numpy as np

# Our sketch generating function
def sketch(image):
    
    img_gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
    img_gray_blur = cv2.GaussianBlur(img_gray, (5,5), 0)
    canny_edges = cv2.Canny(img_gray_blur, 10, 70)
    
    # Do an invert binarize the image 
    ret, mask = cv2.threshold(canny_edges, 70, 255, cv2.THRESH_BINARY_INV)
    return mask

cap = cv2.VideoCapture(0)

while True:
    ret, frame = cap.read()
    cv2.imshow('Our Live Sketcher', sketch(frame))
    if cv2.waitKey(1) == 13: #13 is the Enter Key
        break
        
# Release camera and close windows
cap.release()
cv2.destroyAllWindows()      

# Credits

https://www.udemy.com/course/master-deep-learning-computer-visiontm-cnn-ssd-yolo-gans